gcloud-java-dns/src/main/java/com/google/gcloud/dns/testing/LocalDnsHelper.java

```diff
@@ -0,0 +1,1360 @@
+          return current;
+        }
+      }
+      return null;
+    }
+  }
+
+  static class Response {
+    private final int code;
+    private final String body;
+
+    Response(int code, String body) {
+      this.code = code;
+      this.body = body;
+    }
+
+    int code() {
+      return code;
+    }
+
+    String body() {
+      return body;
+    }
+  }
+
+  private enum Error {
+    REQUIRED(400, "global", "required", "REQUIRED"),
+    INTERNAL_ERROR(500, "global", "internalError", "INTERNAL_ERROR"),
+    BAD_REQUEST(400, "global", "badRequest", "BAD_REQUEST"),
+    INVALID(400, "global", "invalid", "INVALID"),
+    CONTAINER_NOT_EMPTY(400, "global", "containerNotEmpty", "CONTAINER_NOT_EMPTY"),
+    NOT_AVAILABLE(400, "global", "managedZoneDnsNameNotAvailable", "NOT_AVAILABLE"),
+    NOT_FOUND(404, "global", "notFound", "NOT_FOUND"),
+    ALREADY_EXISTS(409, "global", "alreadyExists", "ALREADY_EXISTS"),
+    CONDITION_NOT_MET(412, "global", "conditionNotMet", "CONDITION_NOT_MET"),
+    INVALID_ZONE_APEX(400, "global", "invalidZoneApex", "INVALID_ZONE_APEX");
+
+    private final int code;
+    private final String domain;
+    private final String reason;
+    private final String status;
+
+    Error(int code, String domain, String reason, String status) {
+      this.code = code;
+      this.domain = domain;
+      this.reason = reason;
+      this.status = status;
+    }
+
+    Response response(String message) {
+      try {
+        return new Response(code, toJson(message));
+      } catch (IOException e) {
+        return Error.INTERNAL_ERROR.response("Error when generating JSON error response.");
+      }
+    }
+
+    private String toJson(String message) throws IOException {
+      Map<String, Object> errors = new HashMap<>();
+      errors.put("domain", domain);
+      errors.put("message", message);
+      errors.put("reason", reason);
+      Map<String, Object> args = new HashMap<>();
+      args.put("errors", ImmutableList.of(errors));
+      args.put("code", code);
+      args.put("message", message);
+      args.put("status", status);
+      return jsonFactory.toString(ImmutableMap.of("error", args));
+    }
+  }
+
+  private class RequestHandler implements HttpHandler {
+
+    /**
+     * Chooses the proper handler for a request.
+     */
+    private Response pickHandler(HttpExchange exchange, CallRegex regex) {
+      String path = BASE_CONTEXT.relativize(exchange.getRequestURI()).getPath();
+      String[] tokens = path.split("/");
+      String projectId = tokens != null && tokens.length > 0 ? tokens[0] : null;
+      String zoneName = tokens != null && tokens.length > 2 ? tokens[2] : null;
+      String changeId = tokens != null && tokens.length > 4 ? tokens[4] : null;
+      String query = BASE_CONTEXT.relativize(exchange.getRequestURI()).getQuery();
+      switch (regex) {
+        case CHANGE_GET:
+          return handleChangeGet(projectId, zoneName, changeId, query);
```

Comment:
I'm usually all for factoring methods into multiple pieces, but it seems like these handlers mostly add one line of functionality before passing on this information to the next method (e.g. getChange, listChange, etc).  It be more concise to call those methods directly from here, passing along the query as a parameter to be parsed in the handler method.